### PR TITLE
Provide newer versions of llvm (3.8.1, 3.9.0)

### PR DESCRIPTION
**Pull request by KineticTheory**
+ Provide file list and md5 hashes for 3.8.1 and 3.9.0.
+ Clean up indentation for the 'releases' data structure to improve consistency.

LLVM has a complicated recipe and I'm not an expert on building it.  The changes in this PR worked for me, but I would like a second opinion before they are merged.

### CODE REVIEW

File: var/spack/repos/builtin/packages/llvm/package.py [hunk 1]
```diff
@@ -141,63 +141,93 @@ class Llvm(Package):
...
+            {
+                'version': '3.9.0',
+                'md5': 'f2093e98060532449eb7d2fcfd0bc6c6',
+                'resources': {
+                    'compiler-rt': 'b7ea34c9d744da16ffc0217b6990d095',
+                    'openmp': '5390164f2374e1444e82393541ecf6c7',
+                    'polly': '1cf328cbae25267749b68cfa6f113674',
+                    'libcxx': '0a11efefd864ce6f321194e441f7e569',
+                    'libcxxabi': 'd02642308e22e614af6b061b9b4fedfa',
+                    'cfe': '29e1d86bee422ab5345f5e9fb808d2dc',
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/llvm/package.py]: Should this be cfe or clang? It's failing the unit tests.

**Comment by KineticTheory** [comment]
@adamjstewart for 3.9.0 and 3.8.1 it should be `cfe`.  See http://llvm.org/releases/3.9.0/. I didn't modify the `trunk` revision coding.  If I look at it now, I think it needs to be updated to `cfe` in place of `clang`.  I'll give it a try and then upload a modified `package.py`.

**Comment by KineticTheory** [comment]
@adamjstewart Actually, the build for version `trunk` doesn't seem to work at all, but I did find a bug in my PR that is now fixed in commit 6070603.

**Comment by adamjstewart** [comment]
Ah, that looks much better. Thanks!

**Comment by tgamblin** [comment]
@trws: can you look at this one before I merge it?

**Comment by trws** [comment]
I'll give it a try.  In the meantime, why the change to cfe? Does the directory in the actual build tree need to be named `cfe` now?




On September 21, 2016 at 12:24:43 AM PDT, Todd Gamblin <notifications@github.com> wrote:

@trws<https://github.com/trws>: can you look at this one before I merge it?

-

**Comment by tgamblin** [comment]
@KineticTheory

**Comment by KineticTheory** [comment]
@trws I'm not familiar enough with the llvm build process to provide a definitive answer to your question about the name change.  All I know is that the recipe that I provided produces the tools I wanted to use: clang, clang++ and clang-format (v 3.9). Can you attempt to build version 3.9 with this recipe and recommend improvements?

If you are asking about the `cfe` entry in `resources`,  the string `'placement' : 'clang'` works, but `'placement' : 'cfe'` does not.

**Comment by trws** [comment]
That's what I thought.  I already had a "clang" resource that did the 
same thing, so there's some unnecessary duplication there.  I'm building 
a 3.9 right now to make sure things behave.

**Comment by adamjstewart** [comment]
From what I can tell, your keys are identical, they just have different names. Either will work, but we should settle on one or the other for consistency. Personally I would vote for cfe since all of the other keys are named after the url, not the placement.

**Comment by trws** [comment]
The 3.9 build worked just fine.  As @adamjstewart says, the cfe and clang resource keys should be merged, either way is fine with me really.  Other than that, I'm good with this PR.

**Comment by tgamblin** [comment]
experimenting with the new github review thing.

File: var/spack/repos/builtin/packages/llvm/package.py (hunk 1 above)
Comment (KineticTheory) [inline var/spack/repos/builtin/packages/llvm/package.py]: See commit 00beee9.

**Pull request merged**